fix: make the `virtual_text` diagnostic respect the diagnostic columns

Thibaulltt opened the pull request:
# What does this do ?

This is a proposal to add a simple new parameter to the `vim.diagnostic.Opts.VirtualText` table called `as_inlay`. This parameter allows the `virtual_text` of any given diagnostics to be set at the 'correct' location in a line, and visually displayed as an inlay hint. This parameter can be `true`, `false` or `nil` (default).

If `false`/`nil`, the previously implemented behaviour is enabled: all diagnostics for a line are bunched up at the 0-th column and displayed using an extmark placed at that column.

If `true`, it groups the diagnostics on each line based on the column where they appear, and later produces text chunks based on the diagnostic messages. These chunks are produced using the same logic as before.

# Why ?

Previously, all diagnostics for a line would be grouped together at the user-defined position (based on the `virt_text_pos` parameter) and this would result in all diagnostics being bunched up on the start/end of the line, which is (in my opinion) not desired. If I enable `virtual_text`, I expect my issues to be reported where they happen -- or at least have the option to. This is also supported by the fact the `underline` handler actually highlights the right region.

## Minimal example for the current behaviour

Given the following `minimal.lua` file (this expects the `lua_ls` LSP server to be installed):

```lua
for name, url in pairs {
  ['nvim-lspconfig'] = 'https://github.com/neovim/nvim-lspconfig',
} do
  local install_path = vim.fn.fnamemodify('nvim-issue-diagnostics/' .. name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

-- LSP setup:
local M = {}
M.settings = {
	Lua = {
		runtime = { version = 'LuaJIT', builtin = 'Disabled', },
		diagnostics = { enable = true, globals = {}, },
		workspace = { library = {}, checkThirdParty = 'Disabled', },
		telemetry = { enable = false, }
	},
}
M.filetypes = { 'lua' }
M.root_markers = { ".luarc.json", ".luarc.jsonc", ".luacheckrc", ".stylua.toml", "stylua.toml", "selene.toml", "selene.yml", ".git", "init.lua" }

vim.lsp.config('lua_ls', M)
vim.lsp.enable('lua_ls')

vim.diagnostic.config({
	virtual_text = {
                -- as_inlay = true, -- To add later for the new behaviour!
		spacing = 1,
		virt_text_pos = 'inline',
		hl_mode = 'replace',
	},
	float = { border = "rounded" },
	severity_sort = true,
	update_in_insert = false,
})
```

and the following `error.lua` file (notice the single space in line 3):

```lua
local foo = { 'hello' }
-- The following line consists of a single space:
 
local bar = { foo = fo }
```

The diagnostics returned by the `lua_ls` server would all be bunched up at the start of the line. See the following screenshot:
![neovim-diagnostic-inlay-previous](https://github.com/user-attachments/assets/6d247772-3e15-4a91-adc7-8c416a0b61a8)


## Minimal example for the new behaviour

Take the same `error.lua` file, and add `as_inlay = true,` to the `minimal.lua` file inside the `virtual_text` table, and the diagnostics now look like the following:
![neovim-diagnostic-inlay-next](https://github.com/user-attachments/assets/c81ca93f-d110-47ad-a294-96582eedc232)

# What's next ?

I'm looking for feedback! As this change only concerns the rendering of extmarks, I did not see the need to write new tests (and I also did not see what I could test against!). The changes are fairly minimal, and should not dramatically affect runtime performance.

All tests related to diagnostics pass on my end. I have updated the documentation for the new parameter.

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

Commit:
feat(diagnostics): add as_inlay to virtual_text

This is a new parameter that allows the virtual_text to be set at the
correct location in a line, and displayed as an inlay hint (visually).

Previously, all diagnostics for a line would be grouped together at the
user-defined position (based on the `virt_text_pos` parameter) and this
would result in all diagnostics being bunched up on the start/end of the
line, which is (in my opinion) not desired. If I enable virtual_text, I
expect my issues to be reported where they happen.

This parameter can be true, false or nil (default).

If false/nil, the previously implemented behaviour is enabled: all
diagnostics for a line are bunched up at the 0-th column and displayed
using an extmark placed at that column.

If true, it groups the diagnostics on each line based on the column
where they appear, and later produces text chunks based on the
diagnostic messages. These chunks are produced using the same logic as
before (the `M._get_virt_text_chunks()` function).

There are still tests to write :)

Commit:
docs(diagnostics): add docs for `as_inlay`

glepnir commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1643,6 +1657,42 @@ M.handlers.underline = {
+  if opts.as_inlay then
+    --- Gather diagnostics per-column:
+    for _, diag in ipairs(line_diagnostics) do
+      local col = diag.end_col or diag.col
+      local col_diags = group_diags[col]
+      if col_diags == nil then
+        col_diags = {}
+        group_diags[col] = col_diags
+      end
+      table.insert(col_diags, diag)
```
```suggestion
      local col = diag.end_col or diag.col
      group_diags[col] = group_diags[col] or {}
      table.insert(group_diags[col], diag)
```

Thibaulltt replied:
Ah, I was following the style of `diagnostic_lines()` [as seen here](https://github.com/neovim/neovim/blob/487112d6749d4b7a2b82a60fb30cfb9e21c78394/runtime/lua/vim/diagnostic.lua#L654).

Will change it and push once all tests are finished.

Commit:
docs(diagnostics): document as_inlay in news.txt

Commit:
fix(diagnostics): suggested changes from glepnir

clason commented:
Thank you, but this looks a bit awkward, and I worry about an avalanche of follow-up requests for further configuration (truncation, left or right of position, etc.)

Why does this need to be a built-in handler rather than personal customization (or a small plugin)?

clason commented:
Or am I misunderstanding and this is about the `inline` option not respecting the actual diagnostic (column) position? Wouldn't that just be a bug that needs to be fixed -- why would this need a new option?

clason commented:
> On a somewhat related note: I see multiple references inside the vim.diagnostic runtime file to a cached set of diagnostics (aptly named diagnostics_cache) that seems to be made local to the runtime, and thus unavailable for the plugin developers. In order to reduce the amount of data replicated for any given buffer, can we expect a (maybe read-only) access to the table for third-party plugins ?

Unrelated to this PR, so please open a new issue about this and give _concrete_ examples for the use case and a proposed API. (This is an implementation detail that plugins should not have to worry about, so I am skeptical.)

Commit:
fix(diagnostics): remove as_inlay parameter

- remove the parameter in documentation ;
- remove its uses in the code ;
- remove a LuaLS diagnostic disabling I added by mistake;

Thibaulltt committed:
docs(diagnostics): remove as_inlay parameter

also, remove it from news.txt

Thibaulltt commented:
Alright, it was more of a question than anything, if I really do end up needing it I'll open a separate issue.

Removed the parameter from both `vim.diagnostic.Opts.VirtualText`, its doc and news.txt.

Will update with a new commit once I fix the failing tests, and we should be good to go.

Thibaulltt committed:
fix(diagnostics): ensure col is always valid

fixes the functional tests failing to pass.

Thibaulltt commented:
Alright, all functional and unit tests should be good to go.

Not sure why the `news.txt` test fails, as I removed the parameter from the `news.txt` file and from the documentation. :shrug:

Thibaulltt committed:
fix(diagnostics): fix formatting error
